File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left=0, right=height.size()-1;
        int ans=0;
        int max_left=0, max_right=0;
        while(left<right){
            if(height[left]<height[right]){
                height[left]>=max_left ? (max_left=height[left])
                                        : ans+=(max_left-height[left]);
                ++left;
            }else{
                height[right]>=max_right ? (max_right=height[right])
                                        : ans+=(max_right-height[right]);
                --right;
            }
        }
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n=len(height)
        max_fill=[0]*n
        curr_max=0

        for i in range(1,n-1):
            if height[i-1] > curr_max:
                curr_max=height[i-1]
            max_fill[i]=curr_max

        curr_max=0

        for i in range(n-2,0,-1):
            if height[i+1] > curr_max:
                curr_max=height[i+1]
            max_fill[i]=min(max_fill[i], curr_max)
        
        res=0

        for i in range(1,n-1):
            fill=max_fill[i]-height[i]
            if fill>0:
                res+=fill

        return res